### PR TITLE
(maint) Cache located Position info

### PR DESCRIPTION
**Pull request by justinstoller**
We do not calculate line and column number information during tokenizing
since many tokens do not require that information. Instead we save the
byte offset from the start of the file in each token and the AST node.
Then when we need the location information the "locator" looks up what
line and column would correspond to that byte offset within that file.

This works well if we never save any information between catalog
compiles and have relatively few AST nodes that require there location
information.

However, our usage of location info has expanded over time as well as
users' usage of AST nodes that look up their location info (eg
NamedFunctionExpressions). Now a PE Puppet Server node looks up 11k
locations per compile. None of these are particularly expensive but
amount to a lot of work that we just don't have to do.

There are two important concerns about this:
1. I don't think this is thread safe and consequently not portable to
6.x. I believe we might want to evaluate if it is worth changing the AST
Factory methods to compute their location data when constructed and
dispensing with the locator for most AST Nodes.

2. There appears to be an issue with resource AttributeOperations not
correctly inheriting this behavior. Consequently, this only properly
saves the location data for 7k of the 11k locations used to compile a PE
Puppet Server catalog.

### CODE REVIEW

**Comment by justinstoller** [comment]
The linter didn't like one line of my style. Fixed up.

I've added code to defer the line calculations for Resources and Resource Params until after Evaluation. I'm torn on whether that's a good thing (it should either be good or neutral perf wise).

I also think the issues with AttributeOperations stem from `Puppet::Resource::Param` PCore not descending from `Positioned` but I haven't figured out if I can related them yet.

File: lib/puppet/parser/resource/param.rb [hunk 1]
```diff
@@ -4,14 +4,30 @@ class Puppet::Parser::Resource::Param
...
+  attr_accessor :name, :value, :source, :add, :ast_node
 
   def initialize(hash)
     set_options(hash)
     requiredopts(:name)
     @name = @name.intern
   end
 
+  def line
+    @line ||= @ast_node && @ast_node.line
```
Comment (mwaggett) [inline lib/puppet/parser/resource/param.rb:16]: Can you explain this to me?

Comment (mwaggett) [reply lib/puppet/parser/resource/param.rb:16]: never mind, figured it out lol

**Comment by justinstoller** [comment]
jenkins please test this with servertests

**Pull request closed without merging**